ext/standard/basic_functions.c

```diff
@@ -1131,18 +1131,34 @@ PHP_FUNCTION(flush)
 /* {{{ Delay for a given number of seconds */
 PHP_FUNCTION(sleep)
 {
-	zend_long num;
+	zval *num;
 
 	ZEND_PARSE_PARAMETERS_START(1, 1)
-		Z_PARAM_LONG(num)
+		Z_PARAM_NUMBER(num)
 	ZEND_PARSE_PARAMETERS_END();
-
-	if (num < 0) {
-		zend_argument_value_error(1, "must be greater than or equal to 0");
-		RETURN_THROWS();
-	}
-
-	RETURN_LONG(php_sleep((unsigned int)num));
+    if (Z_TYPE_P(num) == IS_DOUBLE) {
+		const double seconds = Z_DVAL_P(num);
+		if (UNEXPECTED(seconds < 0)) {
+			zend_argument_value_error(1, "must be greater than or equal to 0");
+			RETURN_THROWS();
+		}
+#ifdef HAVE_USLEEP
+		const unsigned int fraction_microseconds = (unsigned int)((seconds - (unsigned int)seconds) * 1000000);
+		if(fraction_microseconds > 0) {
+			usleep(fraction_microseconds);
+		}
+#endif
+		RETURN_LONG(php_sleep((unsigned int)seconds));
+	} else {
+		ZEND_ASSERT(Z_TYPE_P(num) == IS_LONG);
```

Comment:
assert does not make sense here. You need to check long specifically and add `else` for incorrect type.

Reply:
nope, pretty sure Z_PARAM_NUMBER(num) guarantees that it will either be DOUBLE or FLOAT or throw :)
```
$ ./sapi/cli/php -r 'sleep("wtf");'

Fatal error: Uncaught TypeError: sleep(): Argument #1 ($seconds) must be of type int|float, string given in Command line code:1
Stack trace:
#0 Command line code(1): sleep('wtf')
#1 {main}
  thrown in Command line code on line 1
$ time ./sapi/cli/php -r 'sleep("0.1");'

real	0m0.103s
user	0m0.002s
sys	0m0.000s
$ time ./sapi/cli/php -r 'sleep("0.12");'

real	0m0.123s
user	0m0.002s
sys	0m0.000s
```

Reply:
sorry I missed that it's Z_PARAM_NUMBER - thought that it's just zval so should be fine.